pkg/controller/syncmachineset/syncmachineset_controller.go

```diff
@@ -0,0 +1,563 @@
+		return fmt.Errorf("error looking up machineset's cluster: %v", err)
+	}
+	msLog = logging.WithCluster(msLog, coCluster)
+
+	if !ms.Status.ClusterAPIInstalled {
+		msLog.Debugf(errorMsgClusterAPINotInstalled)
+		return fmt.Errorf(errorMsgClusterAPINotInstalled)
+	}
+
+	// Create the Cluster object if it does not already exist:
+	remoteClusterAPIClient, remoteDeploymentClient, err := c.buildClients(ms)
+	if err != nil {
+		return err
+	}
+
+	// handleErr will requeue machine sets after waiting when the
+	// remote cluster API deployment is not ready.
+	remoteClusterAPIDeployment, err := remoteDeploymentClient.AppsV1().Deployments(remoteClusterAPINamespace).Get(remoteClusterAPIDeploymentName, metav1.GetOptions{})
+	if err != nil {
+		return NewErrClusterAPINotInstalled(fmt.Sprintf("unable to retrieve remote cluster API deployment: %v", err))
+	}
+	if remoteClusterAPIDeployment.Status.ReadyReplicas < 1 {
+		return NewErrClusterAPINotInstalled(errorMsgClusterAPINotInstalled)
+	}
+
+	rCluster, err := remoteClusterAPIClient.ClusterV1alpha1().Clusters(remoteClusterAPINamespace).Get(coCluster.Name, metav1.GetOptions{})
+	if errors.IsNotFound(err) {
+		// Cluster does not exist, create it:
+		msLog.Infof("creating new cluster API object in remote cluster")
+		rCluster = buildClusterAPICluster(coCluster)
+		_, err := remoteClusterAPIClient.ClusterV1alpha1().Clusters(remoteClusterAPINamespace).Create(rCluster)
+		if err != nil {
+			return fmt.Errorf("error creating cluster object in remote cluster: %v", err)
+		}
+	} else if err != nil {
+		return err
+	} else {
+		// TODO: patch/update here?
+		msLog.Debugf("cluster API object already exists in remote cluster")
+	}
+
+	return nil
+}
+
+// Creates machinesets in the remote.
+func (c *Controller) syncComputeMachineSet(ms *cov1.MachineSet) error {
+	msLog := logging.WithMachineSet(c.logger, ms)
+	msLog.Debug("loaded compute machineset")
+
+	// Lookup the CO cluster object for this machine set and use
+	// that to get the master machine set for the cluster. Continue
+	// syncing generic machine set if the master machine set is
+	// installed (ms.Status.ClusterAPIInstalled == true).
+	// TODO: Set a flag on cluster object once it has been synced
+	// so that we don't have to retrieve the master machine set to
+	// know if we can sync the other machine sets.
+	coCluster, err := c.clusterLister.Clusters(ms.Namespace).Get(ms.Labels[controller.ClusterNameLabel])
+	if err != nil {
+		return fmt.Errorf("error looking up machineset's cluster: %v", err)
+	}
+	coMasterMachineSet, err := c.machineSetsLister.MachineSets(coCluster.Namespace).Get(coCluster.Status.MasterMachineSetName)
+	if err != nil {
+		return fmt.Errorf("error looking up cluster's master machineset: %v", err)
+	}
+
+	// handleErr will requeue machine sets after waiting when the
+	// remote cluster API deployment is not ready.
+	if !coMasterMachineSet.Status.ClusterAPIInstalled {
+		return NewErrClusterAPINotInstalled(errorMsgClusterAPINotInstalled)
```

Comment:
I am still opposed to the approach of returning an error here instead of enqueueing the compute machine sets when the master machine set is updated. I don't think that this is a situation where the alternative is too much work to justify given the expected short-term nature of this controller.